src/main/java/Checkers/GameState.java

```diff
@@ -0,0 +1,511 @@
+        PiecePosition upLeft = new PiecePosition(position.getColumn() - 1, position.getRow() - 1);
+
+        if (normalPosition.contains(upLeft) || queenPosition.contains(upLeft)) {
+            return upLeft;
+        }
+
+        PiecePosition downLeft = new PiecePosition(position.getColumn() - 1, position.getRow() + 1);
+
+        if (normalPosition.contains(downLeft) || queenPosition.contains(downLeft)) {
+            return downLeft;
+        }
+
+        PiecePosition upRight = new PiecePosition(position.getColumn() + 1, position.getRow() - 1);
+
+        if (normalPosition.contains(upRight) || queenPosition.contains(upRight)) {
+            return upRight;
+        }
+
+        PiecePosition downRight = new PiecePosition(position.getColumn() + 1, position.getRow() + 1);
+
+        if (normalPosition.contains(downRight) || queenPosition.contains(downRight)) {
+            return downRight;
+        }
+
+        return null;
+    }
+
+    private void promote() {
+        possiblePromote.clear();
+        calculatePromote(board.keySet());
+
+        for (PiecePosition position : possiblePromote) {
+            PieceColour pieceColour = getPiece(position);
+
+            if (pieceColour.getPieceColour().isWhite() && pieceColour.getPieceType().isNormal()) {
+                CheckersDrawer.removePiece(position);
+                CheckersDrawer.addPiece(position, new PieceColour(pieceColour.getPieceColour(), PieceColour.Type.QUEEN), false);
+
+                promoteOnBoard(position, pieceColour);
+            }
+
+            if (pieceColour.getPieceColour().isBlack() && pieceColour.getPieceType().isNormal()) {
+                CheckersDrawer.removePiece(position);
+                CheckersDrawer.addPiece(position, new PieceColour(pieceColour.getPieceColour(), PieceColour.Type.QUEEN), false);
+
+                promoteOnBoard(position, pieceColour);
+            }
+        }
+    }
+
+    public void promoteOnBoard(PiecePosition position, PieceColour pieceColour) {
+        removePieceFromBoard(position);
+        addPieceToBoard(position, new PieceColour(pieceColour.getPieceColour(), PieceColour.Type.QUEEN));
+    }
+
+    public void calculatePromote(Set<PiecePosition> positions) {
+
+        Set<PiecePosition> whites = positions.stream()
+                .filter(position -> position.getRow() == 0)
+                .filter(position -> getPiece(position).getPieceColour() == PieceColour.Colour.WHITE)
+                .collect(Collectors.toSet());
+
+        Set<PiecePosition> blacks = positions.stream()
+                .filter(position -> position.getRow() == 7)
+                .filter(position -> getPiece(position).getPieceColour() == PieceColour.Colour.BLACK)
+                .collect(Collectors.toSet());
+
+        possiblePromote.addAll(whites);
+        possiblePromote.addAll(blacks);
+    }
+
+    public void handleMove(PiecePosition position) {
```

Comment:
Aby rozwiązać problem krzyżowej zależności `CheckersDrawer` i `GameState`, moim zdaniem metoda `handleMove` powinna być w `CheckersDrawer`. Powinna ograniczać się do kwestii wizualnych (dodanie pionka do `gridPane`, usunięcie pionka itp.), a samo wyliczanie np. dostępnych pozycji - może pozostać w `GameState`.

Ogólnie klik na pionka powinien - jeśli dobrze rozumiem - po prostu wyświetlić dostępne położenia docelowe. I na tym można się skupić - z `gameState` pobrać możliwe położenia i dodać takie elementy (`light`) do `gridPane`. 

Zauważ, że element docelowy (ta żółta ramka), na który klikasz, może mieć inny handler do obsługi. W tym handlerze do `gameState` można przekazać wybrany ruch (czyli pozycję docelową - pozycja pierwotna powinna być zapamiętana w `gameState`). Następnie pobrać z `gameState` listę pionków do usunięcia (czyli czy coś zostało zbite).

Rozbijając to na 2 różne handlery, znacząco uprościsz sobie logikę.

Podsumowując:
1. Handler od kliku w pionek:
- pobierz listę dopuszczalnych ruchów z `gameState`, zapamiętując w gameState, który pionek został kliknięty (np. pole `clickedPosition`)
- jeśli są jakieś możliwe ruchy dla tego pionka - dodaj tam elementy `light`, każdemu z nich ustawiając `.setOnMouseClicked(event ->...` <- tu wywołanie metody obsługującej drugą część ruchu (handler od kliku w light).

2. Handler od kliku w light:
- wywołaj na `gameState` metodę aktualizującą ruch (np. `move(PiecePosition targetPosition`)
-  metoda ta może zwracać listę zbitych pionków, jeśli ta lista jest niepuste - w `CheckersDrawer` usuwasz te pionki metodą `removePiece`
- ustawiasz nowe położenie przesuwanego pionka za pomocą `addPiece`
- sprawdzasz warunki zwycięstwa
- wywołujesz wykonanie ruchu przez komputer
- ponownie sprawdzasz warunki zwycięstwa
- usuwasz elementy light z gridPane

Jak to już będzie działać, można wrócić do tematu królowej - wtedy w handlerze kliku w light zamiast zwracania z `gameState` listy zbitych pionków, możesz zwrócić obiekt klasy zawierającej taką listę + info o konwersji na królową.